x/liquidity/amm/pool.go

```diff
@@ -70,78 +80,240 @@ func (pool *BasicPool) IsDepleted() bool {
 	return pool.ps.IsZero() || pool.rx.IsZero() || pool.ry.IsZero()
 }
 
-// Deposit returns accepted x and y coin amount and minted pool coin amount
-// when someone deposits x and y coins.
-func (pool *BasicPool) Deposit(x, y sdk.Int) (ax, ay, pc sdk.Int) {
-	// Calculate accepted amount and minting amount.
-	// Note that we take as many coins as possible(by ceiling numbers)
-	// from depositor and mint as little coins as possible.
-
-	utils.SafeMath(func() {
-		rx, ry := pool.rx.ToDec(), pool.ry.ToDec()
-		ps := pool.ps.ToDec()
+// HighestBuyPrice returns the highest buy price of the pool.
+func (pool *BasicPool) HighestBuyPrice() (price sdk.Dec, found bool) {
+	// The highest buy price is actually a bit lower than pool price,
+	// but it's not important for our matching logic.
+	return pool.Price(), true
+}
 
-		// pc = floor(ps * min(x / rx, y / ry))
-		pc = ps.MulTruncate(sdk.MinDec(
-			x.ToDec().QuoTruncate(rx),
-			y.ToDec().QuoTruncate(ry),
-		)).TruncateInt()
+// LowestSellPrice returns the lowest sell price of the pool.
+func (pool *BasicPool) LowestSellPrice() (price sdk.Dec, found bool) {
+	// The lowest sell price is actually a bit higher than the pool price,
+	// but it's not important for our matching logic.
+	return pool.Price(), true
+}
 
-		mintProportion := pc.ToDec().Quo(ps)             // pc / ps
-		ax = rx.Mul(mintProportion).Ceil().TruncateInt() // ceil(rx * mintProportion)
-		ay = ry.Mul(mintProportion).Ceil().TruncateInt() // ceil(ry * mintProportion)
+func (pool *BasicPool) BuyAmountOver(price sdk.Dec, _ bool) (amt sdk.Int) {
+	if price.GTE(pool.Price()) {
+		return sdk.ZeroInt()
+	}
+	dx := pool.rx.ToDec().Sub(price.MulInt(pool.ry)).TruncateInt()
+	if dx.IsZero() {
+		return sdk.ZeroInt()
+	}
+	utils.SafeMath(func() {
+		amt = pool.rx.ToDec().QuoTruncate(price).Sub(pool.ry.ToDec()).TruncateInt()
```

Comment:
What if `price` is too small like `price=0`? Do we need some condition to set `price` as the lowest price?